src/main/java/com/github/retro_game/retro_game/service/impl/ResetPasswordServiceImpl.java

```diff
@@ -0,0 +1,117 @@
+package com.github.retro_game.retro_game.service.impl;
+
+import com.github.retro_game.retro_game.entity.User;
+import com.github.retro_game.retro_game.entity.UserPasswordResetToken;
+import com.github.retro_game.retro_game.entity.UserPasswordResetTokenKey;
+import com.github.retro_game.retro_game.repository.UserPasswordResetTokenRepository;
+import com.github.retro_game.retro_game.repository.UserRepository;
+import com.github.retro_game.retro_game.service.EmailService;
+import com.github.retro_game.retro_game.service.ResetPasswordService;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.apache.commons.lang3.time.DateUtils;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Date;
+import java.util.Optional;
+
+@Service
+public class ResetPasswordServiceImpl implements ResetPasswordService {
+  private final PasswordEncoder passwordEncoder;
+  private final UserRepository userRepository;
+  private final EmailService emailService;
+  private final UserPasswordResetTokenRepository userPasswordResetTokenRepository;
+  private final int passwordResetLinkExpiresAfterDays;
+
+  public ResetPasswordServiceImpl(
+      @Value("${retro-game.password-reset-link-expires-after-days}") int passwordResetLinkExpiresAfterDays,
+      PasswordEncoder passwordEncoder,
+      UserRepository userRepository,
+      EmailService emailService,
+      UserPasswordResetTokenRepository userPasswordResetTokenRepository) {
+    this.passwordEncoder = passwordEncoder;
+    this.userRepository = userRepository;
+    this.emailService = emailService;
+    this.userPasswordResetTokenRepository = userPasswordResetTokenRepository;
+    this.passwordResetLinkExpiresAfterDays = passwordResetLinkExpiresAfterDays;
+  }
+
+  @Override
+  @Transactional
+  public void generateTokenAndSendEmail(String email) {
+    Optional<User> user = userRepository.findByEmailIgnoreCase(email);
+
+    user.ifPresent(u -> {
+      String plainToken = generateToken();
+      Optional<UserPasswordResetToken> existingToken = userPasswordResetTokenRepository.findByKey_User(u);
+
+      if (existingToken.isPresent()) {
+        userPasswordResetTokenRepository.save(
+            updatedToken(plainToken, existingToken.get())
+        );
```

Comment:
Should we also refresh the `expireAt` value if a token already exists for the user? If we don't update it, the user might generate 2nd token after a few days without using the 1st token, but the 2nd token can be already expired when it is generated. In that case, the user will never be able to reset the password.

Reply:
Agree, thanks for hint. Added `exprieAt` update to `updatedToken`.